### ISSUE TITLE
Animation done callback being fired twice on IE11 and Edge

### ISSUE THREAD

**Issue by crisbeto** [angular/angular#24084]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Performance issue
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
[ ] Other... Please describe:
</code></pre>

## Current behavior
Currently having a `(@someAnimation.done)` callback will fire once for most browsers, but twice on IE11 and Edge.

## Expected behavior

The callback should only fire once on all browsers.

## Minimal reproduction of the problem with instructions
The issue can be reproduced on the Angular Material select component. The event is bound in the [select's template](https://github.com/angular/material2/blob/master/src/lib/select/select.html#L38) and [the animation definition is here](https://github.com/angular/material2/blob/master/src/lib/select/select-animations.ts#L40). 

[Here's a Stackblitz from the related issue](https://stackblitz.com/edit/angular-tqzukt) on the Material repo (https://github.com/angular/material2/issues/11444).

**Note:** the callback seems to fire correctly for the `void => *` transitions, but it ends up firing twice for `* => void`.

## Environment

<pre><code>
Angular version: 6.0.1

Browser:
- [ ] Chrome (desktop) version XX
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [x] IE version 11
- [x] Edge version 42

Others:

</code></pre>

**Comment by denisyilmaz** [comment]
same for me in Safari 11.1
every transition is calling done twice. (not in material, but animation triggers in general). (Angular 6.0.4)

**Comment by denisyilmaz** [comment]
Reproduction: https://stackblitz.com/edit/angular-61htna

**Comment by denisyilmaz** [comment]
any updates?

**Comment by hql287** [comment]
I'm also experiencing this on Edge & IE on Windows 10.

**Comment by stalari16** [comment]
I'm also facing this issue on IE and Safari.

**Comment by elron** [comment]
I'm facing this issue on Safari on iOS.

**Comment by elron** [comment]
I come up with an awkward workaround that works in my project, maybe it will help you:


Basically I found that it runs the function's event `$event.phaseName = done` is called twice in Safari and Edge, instead of once.

So I decided to only let done() function run ONLY once, if an animation has already started.
The way I did it:

I initially stored `unfinishedAnimation: boolean = false`, and when the animation is called with `$event.phaseName = start`, it will change the boolean to `true`.

If the animation is finished (by means: `$event.phaseName == 'done'`), then run done() if the `unfinishedAnimation === true`, and set 'unfinishedAnimation' as false at this stage. **that way, the second (expected, but unnecessary) call will prevent done() from being called twice, and it will just ignore it and do nothing.**

```
  unfinishedAnimation: boolean = false; // run 'done' commands only if there's an unfinishedAnimation



  animate($event, $run?) {
    console.debug('event', $event);
    if ($event.phaseName == 'start') {
      this.unfinishedAnimation = true;
    }
    if (this.unfinishedAnimation) {
      if ($event.phaseName == 'done') {
        this.unfinishedAnimation = false;
        done(); // your done event goes here
      }
    } 
    // else do nothing...
  }
```

and in my html file:

```
<div [@popOverState]="stateName" (@popOverState.start)="animate($event)"
      (@popOverState.done)="animate($event)" [style.touch-action]="'pan-y'" (swipeleft)="prevStep()" (swiperight)="nextStep()">
      content that get swiped...
</div>
```

**Comment by 6l3m** [comment]
Hi all,
I faced the same issue on IE11/Edge but was able to work around it through the AnimationPlayer's onDone() method (https://angular.io/api/animations/AnimationPlayer#onDone).

Just build your animation programmatically with the AnimationBuilder (https://angular.io/api/animations/AnimationBuilder#usage-notes), the `player.onDone(fn: () => void)` callback should then be called only once.

**Comment by neildocherty** [comment]
I think I have a related issue.

When I animated a staggered list in Safari it triggers `(@someAnimation.done)` once when the list has multiple items but twice when it has only one.

**Comment by rene-leanix** [comment]
Thanks for workaround, @elron! Here's a minor rewrite of your logic using RxJS:

### Component
```ts
export class Component {
  animationActive$ = new BehaviorSubject<boolean>(false);
  constructor() {
    this.animationActive$.pipe(
      distinctUntilChanged(),
      filter(active => !active)
    ).subscribe(() => {
      // This is called only once now for each animation.
    });
  }
}
```
 
### Template
```html
```

**Comment by jessicajaniuk** [comment]
Closing this as there's an effective workaround and we no longer support Internet Explorer. Feel free to file a new issue if this persists.
